### ISSUE TITLE
Not importing files at top of the file is a PEP8 violation

### ISSUE THREAD

**Issue by DavidHwu** [PyCQA/pycodestyle#264]
not importing files at top is a PEP8 violation.

http://legacy.python.org/dev/peps/pep-0008/
Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Not flagged in current Pycharm... rerouted from Pycharm per their request

**Comment by dubglan** [comment]
@IanLee1521 
This is pretty restrictive. There're cases when you have to do conditional imports. There're libraries which, I know it's terrible, mandate configuration to be propagated through environment variables before import. There're various edge cases. As far as I know, pep8 tool doesn't allow disabling warnings per file like pylint does (# pylint:disable=XXX). Without this feature we have to disable E402 permanently in the project config. Please let me know if there's a better way to work around this.

**Comment by IanLee1521** [comment]
Hi @dubglan -- Thanks for the feedback! I would argue that this is an issue more with the PEP than with the tool. I agree with you that there are code bases (e.g. the Python 3.4 standard library) that violate this rule, and there are definitely cases where you might want to do something else. In that case the tool should follow the PEP (in my opinion) and the user should choose whether that is a rule that they want to follow for their code base, or not.

That said, I think the idea of having something to disable warnings on a per-file basis is a reasonable request and so I'll open that as a separate ticket.

**Comment by Mark-E-Hamilton** [comment]
Sorry I don't follow the pep8 issues. We recently upgraded to pep8-1.6.2 and started getting this violation reported
````
test.py:10:1: E402 module level import not at top of file                                      
````
The reason we get these is that our python scripts have this at the beginning:
```
#!/usr/bin/env python
"""Docstring"""
from __future__ import print_function
import sys
sys.dont_write_bytecode = True
import os
...
````
The reason we do this is this is the only way we know of to prevent the .pyc/.pyo files from being generated. Putting the assignment after the other imports doesn't prevent this. I also like the idea of being able to add per-file (or even per-line) disabling of certain warnings.

Thanks.

**Comment by jayvdb** [comment]
@Mark-E-Hamilton , is "sys.dont_write_bytecode = True" needed for correctness of the algorithms within the module?

The docs suggest cmd arg -B and env var PYTHONDONTWRITEBYTECODE will achieve the same objective.

https://docs.python.org/2/library/sys.html#sys.dont_write_bytecode
https://docs.python.org/3/library/sys.html#sys.dont_write_bytecode

**Comment by Mark-E-Hamilton** [comment]
On 02/25/15 18:47, John Vandenberg wrote:
> @Mark-E-Hamilton <https://github.com/Mark-E-Hamilton> , is
> "sys.dont_write_bytecode = True" needed for correctness of the
> algorithms within the module?
>
> The docs suggest cmd arg -B and env var PYTHONDONTWRITEBYTECODE will
> achieve the same objective.
>
> https://docs.python.org/2/library/sys.html#sys.dont_write_bytecode
> https://docs.python.org/3/library/sys.html#sys.dont_write_bytecode

It is not required for correctness; we chose that approach because it 
encapsulated it within the code. The other two approaches require 
something external to the script to set those values, which we can't 
easily control externally, or on the shebang line.

-- 
----------------
Mark E. Hamilton
Engineering Sciences Center
Senior Member of Technical Staff
Sandia National Laboratories
505-844-7666

**Comment by IanLee1521** [comment]
@Mark-E-Hamilton -- Is it possible to set `sys.dont_write_bytecode` after all of the imports? E.g.:

```
#!/usr/bin/env python
"""Docstring"""
from __future__ import print_function
import sys
import os
sys.dont_write_bytecode = True
...
```

**Comment by Mark-E-Hamilton** [comment]
On 02/26/15 15:08, Ian Lee wrote:
> @Mark-E-Hamilton <https://github.com/Mark-E-Hamilton> -- Is it possible
> to set |sys.dont_write_bytecode| after all of the imports? E.g.:

Yes, but because that variable prevents imports from creating .pyc/.pyo 
files, doing it after the imports is too late.


-- 
----------------
Mark E. Hamilton
Engineering Sciences Center
Senior Member of Technical Staff
Sandia National Laboratories
505-844-7666

**Comment by GFiedler-ADC** [comment]
FWIW, the way this issue is implemented seems to contradict the [Version Bookkeeping](https://www.python.org/dev/peps/pep-0008/#version-bookkeeping) section of PEP 8:

> If you have to have Subversion, CVS, or RCS crud in your source file, do it as follows.
> ```python
> __version__ = "$Revision$"
> # $Source$
> ```
> These lines should be included after the module's docstring, before any other code, separated by a blank line above and below. 

pep8 now complains about such lines if they are before any imports.

**Comment by josch** [comment]
In my case I'm doing

    sys.path.append('/some/path')

before my `import` statement which is now also detected as an error.

**Comment by opoplawski** [comment]
I'm seeing the same thing as GFledler-ADC with things like:
~~~~~
__author__ = "Nilton Volpato"
__author_email__ = "first-name dot last-name @ gmail.com"
__date__ = "2006-05-07"
__version__ = "2.2"
~~~~~
just after the docstring.

**Comment by opoplawski** [comment]
Another case:
~~~~~
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
~~~~~
The use statement must come before the pyplot import.

**Comment by danijar** [comment]
+1

**Comment by styshoo** [comment]
>  josch commented on 23 Sep 2015

In my case I'm doing

sys.path.append('/some/path')

before my import statement which is now also detected as an error.

so, how did you fix this problem?

**Comment by josch** [comment]
@styshoo by using `--ignore=E402`

**Comment by peterjc** [comment]
Assume intention here is that ``__docformat__`` lines should also come after the imports section?

https://www.python.org/dev/peps/pep-0258/ just says *A special variable, ``__docformat__`` , may appear at the top level of a module before any function or class definitions.*

**Comment by sigmavirus24** [comment]
@peterjc why would we observe a [rejected PEP](https://www.python.org/dev/peps/pep-0258/#rejection-notice)?

I agree that this rule is restrictive for some modules. That said, ignoring it globally is the best option for now. When Flake8 3.0 comes out you'll be able to ignore just this specific error inline, but until then, your best bet is to add this to the list of errors/warnings in flake8 that you're ignoring.

Sorry everyone.

**Comment by peterjc** [comment]
@sigmavirus24 We've been using ``__docformat__`` via epydoc, so this is a defacto standard even if PEP258 was rejected (which I hadn't noticed, thanks!).

But yes, E402 is much stricter than the tool used to be.

**Comment by warsaw** [comment]
I'll note that the main place I ran into this is with an __all__ that preceded the import lines.  I think that should be acceptable as __all__ is a declaration of API instead of implementation detail, and it's better being at the top of the file rather than having to hunt around for it.

That being said it was a great impetus for http://public.readthedocs.io/en/latest/ and http://bugs.python.org/issue26632

:)

**Comment by sigmavirus24** [comment]
@warsaw the one potential problem with that is that `__all__` is expected to be a tuple by `pydocstyle` so if people are using that tool, they're going to be warned that it is mutable and cannot be properly analyzed.

@peterjc I had no clue that epydoc was using PEP 258's `__docformat__` for that. Thank you for explaining. At this point I'm fairly convinced this could be off by default for the time being. @IanLee1521 thoughts?

**Comment by warsaw** [comment]
@sigmavirus24 Really?  It has to be a tuple? That seems like an arbitrary restriction.

**Comment by sigmavirus24** [comment]
@warsaw not as far as Python is concerned, but as far as `pydocstyle` is concerned, yes. This was more of a constraint imposed for [simplicity](https://github.com/PyCQA/pydocstyle/issues/62#issuecomment-34850909) as I understand it.

**Comment by rst0git** [comment]
Thanks for the fix, the problem with `sys.path` can be solved by replacing
`sys.path.append('/some/path')`
with:
`sys.path += ['/some/path']`
to ignore the issue with pycodestyle, you can add:
`sys.path += ['/some/path']  # noqa: E402`

**Comment by sigmavirus24** [comment]
@rst0git `# noqa: E402` only works on Flake8 3.0+ and thus isn't relevant to pycodestyle users who don't run it via Flake8. Also, please do not comment on year old issues that have been closed for longer than that.

**Comment by waketzheng** [comment]
> Another case:
> 
> ```
> import matplotlib
> matplotlib.use("Agg")
> import matplotlib.pyplot as plt
> ```
> 
> The use statement must come before the pyplot import.

I fix it by this way:
```py
import contextlib

import matplotlib

matplotlib.use("Agg")

with contextlib.suppress(FutureWarning):  # Ignore future deprecated warnings
    import matplotlib.pyplot as plt
```
